lib/balanced_vrp_clustering.rb

```diff
@@ -49,37 +50,40 @@ def build(data_set, cut_symbol, cut_ratio = 1.0, options = {})
         #    index 3 : unit_quantities -> for each unit, quantity associated to this item
         #    index 4 : characteristics -> { v_id: sticky_vehicle_ids, skills: skills, days: day_skills, matrix_index: matrix_index }
 
-        ### return clean errors if unconsistent data ###
-        if cut_symbol && data_set.data_items.any?{ |item| item[3].nil? || !item[3].has_key?(cut_symbol) }
-          raise ArgumentError, 'Cut symbol corresponding unit should be provided for all item'
+        @vehicles ||= @vehicles_infos # (DEPRECATED)
+
+        ### return clean errors if inconsistent data ###
+        if distance_matrix
+          if @vehicles.any?{ |v_i| v_i[:depot].size != 1 } ||
+             data_set.data_items.any?{ |item| !item[4][:matrix_index] }
+            raise ArgumentError, 'Distance matrix provided: matrix index should be provided for all vehicles and items'
+          end
+        elsif @vehicles.any?{ |v_i| v_i[:depot].compact.size != 2 }
+          raise ArgumentError, 'Location info (lattitude and longitude) should be provided for all vehicles'
         end
 
-        raise ArgumentError, 'Should provide max_iterations' if @max_iterations.nil?
+        if data_set.data_items.any?{ |item| !(item[0] && item[1]) }
+          raise ArgumentError, 'Location info (lattitude and longitude) should be provided for all items'
+        end
 
-        ### default values ###
-        data_set.data_items.each{ |item|
-          item[4][:v_id] ||= []
-          item[4][:skills] ||= []
-          item[4][:days] ||= ['0_day_skill', '1_day_skill', '2_day_skill', '3_day_skill', '4_day_skill', '5_day_skill', '6_day_skill']
-        }
+        if cut_symbol && !@vehicles.all?{ |v_i| v_i[:capacities].has_key?(cut_symbol) }
+          raise ArgumentError, 'All vehicles should have a limit for the unit corresponding to the cut symbol'
+        end
 
-        vehicles_infos.each{ |vehicle_info|
-          vehicle_info[:total_work_days] ||= 1
-          vehicle_info[:skills] ||= []
-          vehicle_info[:days] ||= ['0_day_skill', '1_day_skill', '2_day_skill', '3_day_skill', '4_day_skill', '5_day_skill', '6_day_skill']
-        }
+        if cut_symbol && data_set.data_items.any?{ |item| item[3].nil? || !item[3].has_key?(cut_symbol) }
+          raise ArgumentError, 'The unit corresponding to the cut symbol should be provided for all items'
+        end
 
+        ### values ###
+        @logger = options[:logger]
```

Comment:
I see where you are coming from and I agree; 

I think, eventually, we should be using the `initializer` and `new` for this purpose. Generate a clusterer object with its logger and use it. 

Unfortunately, now, we don't have this logic separation for options/parameters_info but it is a v good idea I think. At the moment there are problem dependent data (i.e., one run only) inside the `parameters_info` and vice-versa. For example, `services` are passed as an argument but `vehicles` are passed as parameters_info. This separation is historical -- i.e., due to the structure of Ai4r. I think we can rework it as soon as we have time.

Reply:
Done.

I just realized that there is no point running vrp clustering multiple times, that functionality is a left-over from Ai4r logic (because it was for data clustering, the idea is, you `build` a "model" with a part of the data_set and then `evaluate` new items (unseen by the model) to test the model without moving the centroids). 

However, in our case this does not really make sense, since the goal is to cluster the services to vehicles or zones, we don't have any gain by not including them in the first model building part and after convergence, if we need to cluster the same data again we need to restart from scratch anways because the centroids are already converged, then we gain nothing by "not creating a new clusterer object".